cubical/kernels/chain.py

```diff
@@ -0,0 +1,235 @@
+| Dimension      | Size |
++================+======+
+| Direction      |   d  |
++----------------+------+
+| Model          |   m  |
++----------------+------+
+| Time           |   t  |
++----------------+------+
+| Time Intervals |   ti |
++----------------+------+
+| Frequency      |   f  |
++----------------+------+
+| Freq Intervals |   fi |
++----------------+------+
+| Antenna        |   a  |
++----------------+------+
+| Correlation    |   c  |
++----------------+------+
+
+"""
+from builtins import range
+
+import numpy as np
+from numba import jit, prange
+
+import cubical.kernels
+from cubical.kernels import generics
+from cubical.kernels import full_complex
+
+use_parallel = True if cubical.kernels.num_omp_threads > 1 else False
+use_cache = cubical.kernels.use_cache
+
+# Allocators are the same as for the 2x2 complex kernel.
+allocate_vis_array = full_complex.allocate_vis_array
+allocate_gain_array = full_complex.allocate_gain_array
+allocate_flag_array = full_complex.allocate_flag_array
+
+@jit(nopython=True, fastmath=True, parallel=use_parallel, cache=use_cache, nogil=True)
+def compute_jh(jh, g, t_int, f_int):
+    """
+    Given J\ :sup:`H` (initially populated with the model array) and gains, computes the non-zero 
+    elements of J\ :sup:`H`. J\ :sup:`H` has full time and frequency resolution - solution intervals
+    are used to correctly associate the gains with the model. The result here contains the useful 
+    elements of J\ :sup:`H` but does not look like the analytic solution. This function is called
+    multiple times during a chain computation. Each call applies a different gain term.   
+
+    Args:
+        jh (np.complex64 or np.complex128):
+            Typed memoryview of J\ :sup:`H` array with dimensions (d, m, t, f, a, a, c, c).
+        g (np.complex64 or np.complex128):
+            Typed memoryview of gain array with dimension (d, ti, fi, a, c, c).
+        t_int (int):
+            Number of time slots per solution interval.
+        f_int (int):
+            Number of frequencies per solution interval.
+    """
+    n_dir = jh.shape[0]
+    n_mod = jh.shape[1]
+    n_tim = jh.shape[2]
+    n_fre = jh.shape[3]
+    n_ant = jh.shape[4]
+    g_dir = g.shape[0]
+
+    all_bls = np.array([[i,j] for i in range(n_ant) for j in range(n_ant) if i!=j])
+    n_bl = all_bls.shape[0]
+
+    broadcast_times = np.array([t//t_int for t in range(n_tim)])
+    broadcast_freqs = np.array([f//f_int for f in range(n_fre)])
+    broadcast_dirs = np.array([d%g_dir for d in range(n_dir)])
+
+    for ibl in prange(n_bl):
+        aa, ab = all_bls[ibl,0], all_bls[ibl,1]
+        for i in range(n_mod):
+            for t in range(n_tim):
+                bt = broadcast_times[t]
+                for f in range(n_fre):
+                    bf = broadcast_freqs[f]
+                    for d in range(n_dir):
+                        bd = broadcast_dirs[d]
+
+                        jh00 = jh[d,i,t,f,aa,ab,0,0]
```

Comment:
Isn't this loop the wrong way around from a cache coherency perspective - i.e. d on the far right instead?

Reply:
When I benchmarked the original cython kernels, I found some really surprising and non-intuitive results w.r.t. loop ordering. Is this order here inherited from cython?

Reply:
Yes - I recreated the ordering you found best with Cython. It seemed best with Numba too. So the underlying memory ordering is consistent with the loop ordering even though the "native" ordering remains [d,m,t,f,a,a,c,c] or a variation thereof.